File: app/src/main/java/org/movieos/feeder/sync/Entry.java

package org.movieos.feeder.sync;

import com.google.gson.annotations.SerializedName;

import java.util.Date;

import io.realm.RealmObject;
import io.realm.annotations.Index;
import io.realm.annotations.PrimaryKey;
import io.realm.annotations.Required;

public class Entry extends RealmObject implements IntegerPrimaryKey {
    @PrimaryKey
    @SerializedName("id")
    int mId;

    @Index
    @SerializedName("feed_id")
    int mFeedId;

    @SerializedName("title")
    String mTitle;
    @SerializedName("url")
    String mUrl;
    @SerializedName("author")
    String mAuthor;
    @SerializedName("content")
    String mContent;
    @SerializedName("summary")
    String mSummary;

    @Required
    @SerializedName("created_at")
    Date mCreatedAt;
    @Required
    @SerializedName("published")
    Date mPublished;

    boolean mUnread;
    boolean mUnreadDirty;
    boolean mStarred;
    boolean mStarredDirty;

    Subscription mSubscription;

    public Entry() {
    }

    public int getId() {
        return mId;
    }

    public int getFeedId() {
        return mFeedId;
    }

    public String getTitle() {
        return mTitle;
    }

    public String getUrl() {
        return mUrl;
    }

    public String getAuthor() {
        return mAuthor;
    }

    public String getContent() {
        return mContent;
    }

    public String getSummary() {
        return mSummary;
    }

    public Date getCreatedAt() {
        return mCreatedAt;
    }

    public Date getPublished() {
        return mPublished;
    }

    public boolean isUnread() {
        return mUnread;
    }

    public boolean isStarred() {
        return mStarred;
    }

    public void setUnread(boolean unread, boolean dirty) {
        mUnread = unread;
        mUnreadDirty = dirty;
    }

    public void setStarred(boolean starred, boolean dirty) {
        mStarred = starred;
        mStarredDirty = dirty;
    }

    public Subscription getSubscription() {
        return mSubscription;
    }

    public void setSubscription(Subscription subscription) {
        mSubscription = subscription;
    }
}

File: app/src/main/java/org/movieos/bluereader/model/Entry.kt
package org.movieos.bluereader.model

import android.arch.persistence.room.Entity
import android.arch.persistence.room.Index
import android.arch.persistence.room.PrimaryKey
import android.content.Context
import com.google.gson.annotations.SerializedName
import org.movieos.bluereader.api.Mercury
import java.util.*

@Entity(
        indices = arrayOf(
                Index("feedId", "published"),
                Index("feedId", "published", "unread"),
                Index("feedId", "published", "starred"),
                Index("published"),
                Index("feedId")
        )
)
data class Entry(
        @PrimaryKey
        @SerializedName("id")
        val id: Int,

        @SerializedName("feed_id")
        val feedId: Int,

        @SerializedName("title")
        val title: String?,

        @SerializedName("url")
        val url: String?,

        @SerializedName("author")
        val author: String?,

        @SerializedName("content")
        val content: String?,

        @SerializedName("summary")
        val summary: String?,

        @SerializedName("created_at")
        val createdAt: Date?,

        @SerializedName("published")
        val published: Date?,

        // these properties are denormalized locally and stored to make the UI state easier to handle
        var unread: Boolean,
        var starred: Boolean
) {

    val excerpt: String
        get() = summary?.substring(0, Math.min(summary.length, 200)) ?: ""

    fun displayAuthor(subscription: Subscription?): String {
        return when {
            subscription == null ->
                ""
            author == null ->
                subscription.title ?: ""
            else ->
                String.format("%s - %s", subscription.title, author)
        }
    }

    fun hasMercuryContent(context: Context): Boolean {
        return Mercury(context).contentFor(this) != null
    }

}
